src/room.cpp

```diff
@@ -0,0 +1,188 @@
+// Copyright 2017 Tunnler Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include "tunnler/room.h"
+
+#include "tunnler/room_message_types.h"
+
+#include "BitStream.h"
+
+/// Maximum number of concurrent connections allowed to this room.
+static const uint32_t MaxConcurrentConnections = 10;
+
+void Room::Create(const std::string& name, const std::string& server_address, uint16_t server_port) {
+    RakNet::SocketDescriptor socket(server_port, server_address.c_str());
+
+    server = RakNet::RakPeerInterface::GetInstance();
+    // TODO(Subv): Allow specifying the maximum number of concurrent connections.
+    server->Startup(MaxConcurrentConnections, &socket, 1);
+    server->SetMaximumIncomingConnections(MaxConcurrentConnections);
+
+    state = State::Open;
+
+    // Start a network thread to Receive packets in a loop.
+    room_thread = std::make_unique<std::thread>(&Room::ServerLoop, this);
+}
+
+void Room::Destroy() {
+    state = State::Closed;
+
+    {
+        std::lock_guard<std::mutex> lock(server_mutex);
+
+        if (server) {
+            server->Shutdown(300);
+            RakNet::RakPeerInterface::DestroyInstance(server);
+        }
+        server = nullptr;
+    }
+
+    room_thread->join();
+}
+
+void Room::HandleJoinRequest(const RakNet::Packet* packet) {
+    RakNet::BitStream stream(packet->data, packet->length, false);
+
+    stream.IgnoreBytes(sizeof(RakNet::MessageID));
+
+    RakNet::RakString nick;
+    stream.Read(nick);
+
+    MacAddress preferred_mac;
+    stream.Read(preferred_mac);
+
+    // Verify if the nick is already taken.
+    const std::string nickname(nick.C_String(), nick.GetLength());
+    if (!IsValidNickname(nickname)) {
+        SendNameCollision(packet->systemAddress);
+        return;
+    }
+
+    if (preferred_mac != NoPreferredMac) {
+        // Verify if the preferred mac is available
+        if (!IsValidMacAddress(preferred_mac)) {
+            SendMacCollision(packet->systemAddress);
+            return;
+        }
+    } else {
+        // Assign a MAC address of this client automatically
+        preferred_mac = GenerateMacAddress();
+    }
+
+    // At this point the client is ready to be added to the room.
+    Member member{};
+    member.mac_address = preferred_mac;
+    member.nickname = nickname;
+    member.network_address = packet->systemAddress;
+
+    members.push_back(member);
+
+    // Notify everyone that the room information has changed.
+    BroadcastRoomInformation();
+
+    SendJoinSuccess(member);
+}
+
+void Room::ServerLoop() {
+    while (state != State::Closed) {
+        std::lock_guard<std::mutex> lock(server_mutex);
```

Comment:
It might work for his platform.. - This is practically a busyloop.
We should fix this. At least we should have a reminder about it in an issue or something.